### PR TITLE
Misc Fixes 7

### PR DESCRIPTION
**Pull request by raziel1000**
- save_data and app_data fixes
- added some pthread funcs
- added game title to sdl window

### CODE REVIEW

File: src/core/libraries/kernel/memory_management.cpp [hunk 1]
```diff
@@ -225,4 +225,42 @@ int PS4_SYSV_ABI sceKernelGetDirectMemoryType(u64 addr, int* directMemoryTypeOut
...
+                Kernel_Vmm,
+                "BatchMap: entry = {}, operation = {}, len = {:#x}, offset = {:#x}, type = {}, "
+                "result = {}",
+                i, entries[i].operation, entries[i].length, entries[i].offset, (u8)entries[i].type,
+                result);
+
+            if (result == 0)
+                processed++;
+        } else {
+            LOG_ERROR(Kernel_Vmm, "called: Unimplemented Operation = {}", entries[i].operation);
```
Comment (raphaelthegreat) [inline src/core/libraries/kernel/memory_management.cpp]: This should be UNREACHABLE() so we crash on unimplemented opearations

File: src/core/libraries/kernel/thread_management.cpp [hunk 2]
```diff
@@ -1325,6 +1325,10 @@ int PS4_SYSV_ABI posix_sem_post(sem_t* sem) {
     return sem_post(sem);
 }
 
+int PS4_SYSV_ABI posix_sem_destroy(sem_t* sem) {
+    return sem_destroy(sem);
+}
+
```
Comment (raphaelthegreat) [inline src/core/libraries/kernel/thread_management.cpp:1355]: Can you also add sem_timedwait

File: src/emulator.cpp [hunk 3]
```diff
@@ -92,6 +91,11 @@ void Emulator::Run(const std::filesystem::path& file) {
             }
         }
     }
+    std::string game_title = id + " - " + title + " <" + app_version + ">";
```
Comment (raphaelthegreat) [inline src/emulator.cpp]: Can use fmt::format here

File: src/sdl_window.cpp [hunk 4]
```diff
@@ -18,14 +18,15 @@
...
+WindowSDL::WindowSDL(s32 width_, s32 height_, Input::GameController* controller_,
+                     std::string game_title_)
+    : width{width_}, height{height_}, controller{controller_}, game_title{game_title_} {
     if (SDL_Init(SDL_INIT_VIDEO) < 0) {
         UNREACHABLE_MSG("Failed to initialize SDL video subsystem: {}", SDL_GetError());
     }
     SDL_InitSubSystem(SDL_INIT_AUDIO);
 
-    const std::string title = "shadPS4 v" + std::string(Common::VERSION);
+    const std::string title = "shadPS4 v" + std::string(Common::VERSION) + " | " + game_title;
```
Comment (raphaelthegreat) [inline src/sdl_window.cpp]: Here fmt::format too

File: src/sdl_window.h [hunk 5]
```diff
@@ -40,7 +41,8 @@ struct WindowSystemInfo {
 
 class WindowSDL {
 public:
-    explicit WindowSDL(s32 width, s32 height, Input::GameController* controller);
+    explicit WindowSDL(s32 width, s32 height, Input::GameController* controller,
+                       std::string game_title);
```
Comment (raphaelthegreat) [inline src/sdl_window.h]:
```suggestion
                       std::string_view game_title);
```

File: src/sdl_window.h [hunk 6]
```diff
@@ -68,6 +70,7 @@ class WindowSDL {
 private:
     s32 width;
     s32 height;
+    std::string game_title;
```
Comment (raphaelthegreat) [inline src/sdl_window.h]: Game title is only used in constructor so does it need to be a member?

File: src/core/libraries/kernel/memory_management.cpp [hunk 7]
```diff
@@ -225,4 +225,42 @@ int PS4_SYSV_ABI sceKernelGetDirectMemoryType(u64 addr, int* directMemoryTypeOut
...
+                                    int* numEntriesOut, int flags) {
+    int processed = 0;
+    int result = 0;
+    for (int i = 0; i < numEntries; i++) {
+        if (entries == nullptr || entries[i].length == 0 || entries[i].operation > 4) {
+            result = ORBIS_KERNEL_ERROR_EINVAL;
+            break; // break and assign a value to numEntriesOut.
+        }
+
+        if (entries[i].operation == 0) { // MAP_DIRECT
```
Comment (raphaelthegreat) [inline src/core/libraries/kernel/memory_management.cpp]: Might be good idea to make an enum for the operations

**Comment by georgemoralis** [comment]
@squidbus seems like sem_timedwait not supporting on macOS?

File: src/core/libraries/kernel/memory_management.h [hunk 8]
```diff
@@ -31,6 +31,14 @@ enum MemoryProtection : u32 {
     SCE_KERNEL_PROT_GPU_RW = 0x30     // Permit reads/writes from the GPU
 };
 
+enum MemoryOpTypes : u32 {
+    SCE_KERNEL_MAP_OP_MAP_DIRECT = 0,
+    SCE_KERNEL_MAP_OP_UNMAP = 1,
+    SCE_KERNEL_MAP_OP_PROTECT = 2,
+    SCE_KERNEL_MAP_OP_MAP_FLEXIBLE = 3,
+    SCE_KERNEL_MAP_OP_TYPE_PROTECT = 4
```
Comment (georgemoralis) [inline src/core/libraries/kernel/memory_management.h]: ORBIS_

**Comment by squidbus** [comment]
> @squidbus seems like sem_timedwait not supporting on macOS?

There's a similar polyfill I added for missing `pthread_mutex_timedlock`, you can probably just ifdef it out for now and I'll take a look when I can.

**Pull request merged**